### PR TITLE
Break up `RxCtx` into `Ctx.Owner` and `Ctx.Data`

### PR DESCRIPTION
**Pull request by lihaoyi**
This clears up a bunch of confusion, both in our heads and in the code, and also tightens up some guarantees. In reality, the data-dependency-graph and owner-dependency-graph are more-or-less unrelated, so overloading the `RxCtx` to serve the same context confuses everyone. I also broke up `AdvancedTests` because it was too big.

This overall makes things safer: this is code which used to compile fine and fail

```scala
scala> def foo()(implicit ctx: rx.RxCtx) = {
     |   val a = rx.Var(1)
     |   val b = rx.Rx(a() + 1)
     |   println(b())
     |   a
     | }
foo: ()(implicit ctx: rx.RxCtx)rx.Var[Int]

scala> val x = rx.Rx{val y = foo(); y() = y() + 1; println("done!") }
StackOverflowError
```

The problem was that the same `RxCtx` that was being used to transmit "ownership" information was accidentally being used to transmit "data-dependency" information, leading to an accidental dependency-cycle between `a` `b` and `x` and then a stack-overflow. 

With this diff fails to compile

```scala
        def foo()(implicit ctx: rx.Ctx.Owner) = {
          val a = rx.Var(1)
          val b = rx.Rx(a() + 1)
          println(b())
          a
        }


        val x = rx.Rx.unsafe{
          val y = foo(); y() = y() + 1; println("done!")
        }

[error] /Users/haoyi/Dropbox (Personal)/Workspace/scala.rx/scalarx/shared/src/test/scala/rx/AdvancedTests.scala:315: No implicit Ctx.Data is available here!
[error]           println(b())
[error]                    ^
```

Obviously, if you really wanted you could pass an implicit `Ctx.Data` into that function in addition to the `Ctx.Owner`, but unless you have to purposely screw up like that to get the same StackOverflow

Review by @Voltir. Still a first-attempt, but I think it's in relatively good shape.

### CODE REVIEW

File: scalarx/jvm/src/test/scala/rx/EventedTests.scala [hunk 1]
```diff
@@ -1,161 +1,161 @@
...
-      eventually(count == 2)
-      eventually(count == 4)
-      timer.kill()
-      intercept[AssertionError] {
-        eventually(count == 6)
-      }
-    }
-  }
-}
+//package rx
```
Comment (lihaoyi) [inline scalarx/jvm/src/test/scala/rx/EventedTests.scala]: Ooops lemme clean this up

**Comment by lihaoyi** [comment]
Look like I've inadvertently broken 2.10.x compatibility, need to change all the `c.Tree`s back to `c.Expr` and shim `compileTimeOnly`...

**Comment by Voltir** [comment]
For the 2.10 problem, want to give the macro-compat library a shot?

File: scalarx/shared/src/main/scala/rx/Core.scala [hunk 2]
```diff
@@ -290,36 +294,54 @@ class Rx[+T](func: RxCtx => T, owner: Option[RxCtx]) extends Node[T] { self =>
...
-object RxCtx {
-//  @compileTimeOnly("No implicit RxCtx is available here!")
-  object Unsafe extends RxCtx(throw new Exception(
-    "Invalid RxCtx: you can only call `Rx.apply` within an " +
-      "`Rx{...}` block or where an implicit `RxCtx` is available"
-  ))
+object Ctx{
 
-  def safe(): RxCtx = macro Utils.buildSafeCtx
+  object Data extends Generic[Data]{
```
Comment (lihaoyi) [inline scalarx/shared/src/main/scala/rx/Core.scala]: I'm sure there's a way to remove the duplication from here but I haven't figured out how yet

**Pull request merged**